### PR TITLE
[external] add new segement destination for rokt custom audiences

### PR DESCRIPTION
**Pull request by aval-rokt**
_A summary of your pull request, including the what change you're making and why._

We want to allow ROKT customers ( currently only one ) to send attribution events to ROKT using segment platform & our custom action destination. We would like that batching is used where felt required by segment infrastructure.

- We have added a new destination 'rokt-audiences'
- We are using 'CUSTOM' authentication using API_KEY. So for authentication client needs to provide a valid API_KEY ( ROKT provided ), this is a mandatory field used in auth header for invoking ROKT APIs
- We currently only have one _action_ 'upsertCustomAudiences' for this destination. This action will subscribe to 'track' & 'identify' segment events. This action makes a POST request to ROKT API import suppression endpoint to include/exclude emails to ROKT custom audience list.
- We expect below fields to be present as the part of event payload
    - .context.personas.computation_class ( This always needs to be 'audience' for handling Engage->Audience )
    - .context.personas.computation_key ( name of the audience )
    - .context.traits.email OR .traits.email ( user email )
    - .properties OR .traits 
 - We have support for both single & batched events, we parse through the list of events and build a list of (emails & action) to send to ROKT via API

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
Thanks for raising this PR @aval-rokt . I'll schedule it for review.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @aval-rokt , 

A few minor items to fix and one major item. 
Please see inline comments. 

The major item is that batches can contain identify() and track() calls for different Audiences. This is a very common scenario. If you just select the first Audience in the batch and ignore others then there will be data loss and the Audiences in Rokt will be missing data.  

Also, the unit tests are failing. You can run these using the following command: 

yarn jest --testPathPattern=./packages/destination-actions/src/destinations/rokt-audiences

I'll paste in the output from the tests in a comment. 

Kind regards,
Joe

File: packages/destination-actions/src/destinations/rokt-audiences/index.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    testAuthentication: async (request, { settings }) => {
+      const res = await request(CONSTANTS.ROKT_API_BASE_URL + CONSTANTS.ROKT_API_AUTH_ENDPOINT, {
+        method: 'GET',
+        headers: {
+          Authorization: `${settings.apiKey}`
+        }
+      })
+
+      if (res.status !== 200) throw new RetryableError(`Rokt API key authentication check failed`)
```
[the remark this answers is not in the document]
Comment (aval-rokt) [reply packages/destination-actions/src/destinations/rokt-audiences/index.ts]: done

File: packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/index.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+
+import { processPayload } from './custom-audience-operations'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'ROKT build custom audience',
```
[the remark this answers is not in the document]
Comment (aval-rokt) [reply packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/index.ts]: done

File: packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/custom-audience-operations.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    if (!custom_audience_list) {
+      // set audience name to be processed for sending to rokt
+      custom_audience_list = p.custom_audience_name
+    }
+    if (p.custom_audience_name != custom_audience_list) {
+      // this is edge case where a batch contains more than 1 audiences
+      // for the 1st iteration we don't to handle the case of multiple audiences in same batch
+      // so we have decided to ignore other audience name in the batch, once we have picked one to process
+      continue
+    }
```
[the remark this answers is not in the document]
Comment (aval-rokt) [reply packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/custom-audience-operations.ts]: done

File: packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/custom-audience-operations.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (p.traits_or_props[custom_audience_list]) {
+      // audience entered true, include email to list
+      action_map.get(CONSTANTS.INCLUDE)?.push(p.email)
+    } else {
+      // exclude email from list
+      action_map.get(CONSTANTS.EXCLUDE)?.push(p.email)
+    }
```
[the remark this answers is not in the document]
Comment (aval-rokt) [reply packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/custom-audience-operations.ts]: done

**Comment by aval-rokt** [comment]
> Here's the output for failing unit tests:
> 
> ```
> 
> ➜  action-destinations git:(main) yarn jest --testPathPattern=./packages/destination-actions/src/destinations/rokt-audiences
> yarn run v1.22.19
> $ /Users/jayoub/dev/src/github.com/ROKT/action-destinations/node_modules/.bin/jest --testPathPattern=./packages/destination-actions/src/destinations/rokt-audiences
>  PASS  packages/destination-actions/src/destinations/rokt-audiences/__tests__/index.test.ts
>  FAIL  packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/__tests__/index.test.ts
>   ● RoktAudiences.upsertCustomAudiences › should throw an error if the audience creation failed, bad body
> 
>     expect(received).rejects.toThrowError()
> 
>     Received promise resolved instead of rejected
>     Resolved to value: [{"content": "", "data": "", "options": {"afterResponse": [[Function prepareResponse], [Function prepareHeaders], [Function bound afterResponse]], "beforeRequest": [[Function addBasicAuthHeader]], "body": "{\"list\":\"aval_test_two_track_only\",\"action\":\"include\",\"emails\":[\"test.email@test.com\"]}", "headers": {Symbol(map): {"user-agent": ["Segment (Actions)"]}}, "json": {"action": "include", "emails": ["test.email@test.com"], "list": "aval_test_two_track_only"}, "method": "POST", "signal": {}, "statsContext": {}, "throwHttpErrors": true, "timeout": 10000}, "request": {"agent": undefined, "compress": true, "counter": 0, "follow": 20, "size": 0, "timeout": 10000, Symbol(Body internals): {"body": {"data": [123, 34, 108, 105, 115, 116, 34, 58, 34, 97, 118, 97, 108, 95, 116, 101, 115, 116, 95, 116, 119, 111, 95, 116, 114, 97, 99, 107, 95, 111, 110, 108, 121, 34, 44, 34, 97, 99, 116, 105, 111, 110, 34, 58, 34, 105, 110, 99, 108, 117, 100, 101, 34, 44, 34, 101, 109, 97, 105, 108, 115, 34, 58, 91, 34, 116, 101, 115, 116, 46, 101, 109, 97, 105, 108, 64, 116, 101, 115, 116, 46, 99, 111, 109, 34, 93, 125], "type": "Buffer"}, "disturbed": false, "error": null}, Symbol(Request internals): {"headers": {Symbol(map): {"content-type": ["application/json"], "user-agent": ["Segment (Actions)"]}}, "method": "POST", "parsedURL": {"auth": null, "hash": null, "host": "data.stage.rokt.com", "hostname": "data.stage.rokt.com", "href": "https://data.stage.rokt.com/api/1.0/import/suppression", "path": "/api/1.0/import/suppression", "pathname": "/api/1.0/import/suppression", "port": null, "protocol": "https:", "query": null, "search": null, "slashes": true}, "redirect": "follow", "signal": {}}}, "size": 0, "timeout": 10000, Symbol(Body internals): {"body": {"_events": {"prefinish": [Function prefinish]}, "_eventsCount": 1, "_maxListeners": undefined, "_readableState": {"autoDestroy": true, "awaitDrainWriters": null, "buffer": {"head": null, "length": 0, "tail": null}, "closeEmitted": false, "closed": false, "constructed": true, "dataEmitted": false, "decoder": null, "defaultEncoding": "utf8", "destroyed": false, "emitClose": true, "emittedReadable": false, "encoding": null, "endEmitted": false, "ended": true, "errorEmitted": false, "errored": null, "flowing": null, "highWaterMark": 16384, "length": 0, "multiAwaitDrain": false, "needReadable": false, "objectMode": false, "pipes": [], "readableListening": false, "reading": false, "readingMore": false, "resumeScheduled": false, "sync": false, Symbol(kPaused): null}, "_writableState": {"afterWriteTickInfo": null, "allBuffers": true, "allNoop": true, "autoDestroy": true, "bufferProcessing": false, "buffered": [], "bufferedIndex": 0, "closeEmitted": false, "closed": false, "constructed": true, "corked": 0, "decodeStrings": true, "defaultEncoding": "utf8", "destroyed": false, "emitClose": true, "ended": true, "ending": true, "errorEmitted": false, "errored": null, "finalCalled": true, "finished": true, "highWaterMark": 16384, "length": 0, "needDrain": false, "objectMode": false, "onwrite": [Function bound onwrite], "pendingcb": 0, "prefinished": true, "sync": false, "writecb": null, "writelen": 0, "writing": false, Symbol(kOnFinished): []}, "allowHalfOpen": true, Symbol(kCapture): false, Symbol(kCallback): null}, "disturbed": false, "error": null}, Symbol(Response internals): {"counter": 0, "headers": {}, "status": 201, "statusText": "Created", "url": "https://data.stage.rokt.com/api/1.0/import/suppression"}}]
> 
>       60 |     nock(CONSTANTS.ROKT_API_BASE_URL).persist().post(CONSTANTS.ROKT_API_CUSTOM_AUDIENCE_ENDPOINT).reply(400)
>       61 |
>     > 62 |     await expect(
>          |           ^
>       63 |       testDestination.testAction('upsertCustomAudiences', {
>       64 |         event: goodEvent,
>       65 |         useDefaultMappings: true
> 
>       at expect (node_modules/expect/build/index.js:178:15)
>       at Object.<anonymous> (packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/__tests__/index.test.ts:62:11)
> 
>   ● RoktAudiences.upsertCustomAudiences › should throw an error if the audience creation failed, wrong api key auth
> 
>     expect(received).rejects.toThrowError()
> 
>     Received promise resolved instead of rejected
>     Resolved to value: [{"content": "", "data": "", "options": {"afterResponse": [[Function prepareResponse], [Function prepareHeaders], [Function bound afterResponse]], "beforeRequest": [[Function addBasicAuthHeader]], "body": "{\"list\":\"aval_test_two_track_only\",\"action\":\"include\",\"emails\":[\"test.email@test.com\"]}", "headers": {Symbol(map): {"user-agent": ["Segment (Actions)"]}}, "json": {"action": "include", "emails": ["test.email@test.com"], "list": "aval_test_two_track_only"}, "method": "POST", "signal": {}, "statsContext": {}, "throwHttpErrors": true, "timeout": 10000}, "request": {"agent": undefined, "compress": true, "counter": 0, "follow": 20, "size": 0, "timeout": 10000, Symbol(Body internals): {"body": {"data": [123, 34, 108, 105, 115, 116, 34, 58, 34, 97, 118, 97, 108, 95, 116, 101, 115, 116, 95, 116, 119, 111, 95, 116, 114, 97, 99, 107, 95, 111, 110, 108, 121, 34, 44, 34, 97, 99, 116, 105, 111, 110, 34, 58, 34, 105, 110, 99, 108, 117, 100, 101, 34, 44, 34, 101, 109, 97, 105, 108, 115, 34, 58, 91, 34, 116, 101, 115, 116, 46, 101, 109, 97, 105, 108, 64, 116, 101, 115, 116, 46, 99, 111, 109, 34, 93, 125], "type": "Buffer"}, "disturbed": false, "error": null}, Symbol(Request internals): {"headers": {Symbol(map): {"content-type": ["application/json"], "user-agent": ["Segment (Actions)"]}}, "method": "POST", "parsedURL": {"auth": null, "hash": null, "host": "data.stage.rokt.com", "hostname": "data.stage.rokt.com", "href": "https://data.stage.rokt.com/api/1.0/import/suppression", "path": "/api/1.0/import/suppression", "pathname": "/api/1.0/import/suppression", "port": null, "protocol": "https:", "query": null, "search": null, "slashes": true}, "redirect": "follow", "signal": {}}}, "size": 0, "timeout": 10000, Symbol(Body internals): {"body": {"_events": {"prefinish": [Function prefinish]}, "_eventsCount": 1, "_maxListeners": undefined, "_readableState": {"autoDestroy": true, "awaitDrainWriters": null, "buffer": {"head": null, "length": 0, "tail": null}, "closeEmitted": false, "closed": false, "constructed": true, "dataEmitted": false, "decoder": null, "defaultEncoding": "utf8", "destroyed": false, "emitClose": true, "emittedReadable": false, "encoding": null, "endEmitted": false, "ended": true, "errorEmitted": false, "errored": null, "flowing": null, "highWaterMark": 16384, "length": 0, "multiAwaitDrain": false, "needReadable": false, "objectMode": false, "pipes": [], "readableListening": false, "reading": false, "readingMore": false, "resumeScheduled": false, "sync": false, Symbol(kPaused): null}, "_writableState": {"afterWriteTickInfo": null, "allBuffers": true, "allNoop": true, "autoDestroy": true, "bufferProcessing": false, "buffered": [], "bufferedIndex": 0, "closeEmitted": false, "closed": false, "constructed": true, "corked": 0, "decodeStrings": true, "defaultEncoding": "utf8", "destroyed": false, "emitClose": true, "ended": true, "ending": true, "errorEmitted": false, "errored": null, "finalCalled": true, "finished": true, "highWaterMark": 16384, "length": 0, "needDrain": false, "objectMode": false, "onwrite": [Function bound onwrite], "pendingcb": 0, "prefinished": true, "sync": false, "writecb": null, "writelen": 0, "writing": false, Symbol(kOnFinished): []}, "allowHalfOpen": true, Symbol(kCapture): false, Symbol(kCallback): null}, "disturbed": false, "error": null}, Symbol(Response internals): {"counter": 0, "headers": {}, "status": 201, "statusText": "Created", "url": "https://data.stage.rokt.com/api/1.0/import/suppression"}}]
> 
>       71 |     nock(CONSTANTS.ROKT_API_BASE_URL).persist().post(CONSTANTS.ROKT_API_CUSTOM_AUDIENCE_ENDPOINT).reply(401)
>       72 |
>     > 73 |     await expect(
>          |           ^
>       74 |       testDestination.testAction('upsertCustomAudiences', {
>       75 |         event: goodEvent,
>       76 |         useDefaultMappings: true
> 
>       at expect (node_modules/expect/build/index.js:178:15)
>       at Object.<anonymous> (packages/destination-actions/src/destinations/rokt-audiences/upsertCustomAudiences/__tests__/index.test.ts:73:11)
> 
> Test Suites: 1 failed, 1 passed, 2 total
> Tests:       2 failed, 4 passed, 6 total
> Snapshots:   0 total
> Time:        1.996 s
> Ran all test suites matching /.\/packages\/destination-actions\/src\/destinations\/rokt-audiences/i.
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> ```

fixed

File: packages/destination-actions/src/destinations/rokt-audiences/index.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import upsertCustomAudiences from './upsertCustomAudiences'
+import { CONSTANTS } from './constants'
+
+const destination: DestinationDefinition<Settings> = {
+  name: 'Rokt Audiences (Actions)',
+  slug: 'actions-rokt-audiences',
+  mode: 'cloud',
+  description: `
+  This destination allows user to engage audiences using Rokt public API.
+  User can select actions-rokt-audiences as destination when creating engage->audience in segment,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/rokt-audiences/index.ts:14]:
```suggestion
  User can connect Rokt Audiences (Actions) as a destination to their Engage Audience in segment,
```

File: packages/destination-actions/src/destinations/rokt-audiences/index.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  if they intend to create/update custom audiences in Rokt data platform.
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/rokt-audiences/index.ts:15]:
```suggestion
  which will create/update custom audiences in the Rokt data platform.  
```

File: packages/destination-actions/src/destinations/rokt-audiences/index.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+      if (res.status !== 200) throw new InvalidAuthenticationError('Authentication using Rokt ApiKey failed')
+    }
+  },
+
+  extendRequest({ settings }) {
+    return {
+      headers: { Authorization: `${settings.apiKey}` }
+    }
+  },
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/rokt-audiences/index.ts:44]: One last thing that is needed - which I missed from the original review. You'll need to include a preset mapping, or else the customer will have extra setup steps after then create an instance of the Destination. 

In this case it's pretty simple - I'll just give you the code below to include: 

```
presets: [
    {
      name: 'Sync Engage Audience to Rokt',
      subscribe: 'type = "track" or type = "identify"',
      partnerAction: 'upsertCustomAudiences',
      mapping: defaultValues(upsertCustomAudiences.fields)
    }
  ],
```

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @aval-rokt, 
The code changes look good. Test pass. 
Only a couple minor things and we're done. 

Some grammar corrections, and then adding a 'preset'. Please see the comments.

File: packages/destination-actions/src/destinations/rokt-audiences/index.ts (hunk 7 above)
Comment (aval-rokt) [reply packages/destination-actions/src/destinations/rokt-audiences/index.ts:44]: done, thanks for pointing this

**Review by joe-ayoub-segment — APPROVED** [review]
Looks good. 
Feedback from PR implemented. 
Unit tests pass. 
Validation passes.

**Comment by joe-ayoub-segment** [comment]
Hi @aval-rokt , PR merged. It will go out with our next deploy. We haven't schedule a deploy for the moment, but I'll message you as soon as we know the schedule. 
Regards,
Joe

**Comment by aval-rokt** [comment]
> Hi @aval-rokt , PR merged. It will go out with our next deploy. We haven't schedule a deploy for the moment, but I'll message you as soon as we know the schedule. Regards, Joe

Thanks @joe-ayoub-segment , will wait for the deployment. Jus wanted to know that we will be in private beta after this deployment or some other process needs to be completed ?

**Comment by joe-ayoub-segment** [comment]
Hi @aval-rokt , once deployed your Integration will be in Private Beta, and you'll be able to test it in Production. 
I'll send you an email to explain the steps.

**Pull request merged**